File: Sources/BLE/MockProgressorClient.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation

/// A synthetic Progressor.
///
/// The only way to run the app without a physical gauge — **the Simulator has no
/// Bluetooth stack at all** — and what demo mode uses, so someone without hardware
/// (an App Store reviewer included) can see a whole session run.
///
/// Same `@MainActor` shape as the live client; the sample pump is a `Task` created
/// in main-actor context, so it inherits that isolation.
@MainActor
final class MockProgressorClient: ProgressorClient {
    var onEvent: ((ProgressorEvent) -> Void)?
    var onPacketBoundary: ((PacketBoundary) -> Void)?
    var onStateChange: ((ProgressorConnectionState) -> Void)?
    var onDiagnostic: ((ProgressorClientDiagnostic) -> Void)?

    private(set) var state: ProgressorConnectionState = .idle {
        didSet { if state != oldValue { onStateChange?(state) } }
    }
    private(set) var deviceName: String? = "Progressor_MOCK"

    var profile: MockForceProfile

    /// The device streams 80 samples/sec, delivered in batches of ~8.
    private static let sampleHz: Double = 80
    private static let batchSize = 8
    private static let microsPerSample: UInt32 = UInt32(1_000_000 / 80)
    /// `-mockClumpMS N` (DEBUG only): hand batches over in CLUMPS N ms apart instead of
    /// one every 100 ms — an iPad-style Bluetooth stack delivering notifications late and
    /// together. Device timestamps are untouched; only arrival bunches.
    private static let clumpMS: Int = {
        #if DEBUG
        let args = ProcessInfo.processInfo.arguments
        if let flag = args.firstIndex(of: "-mockClumpMS"), flag + 1 < args.count,
           let ms = Int(args[flag + 1]) {
            return ms
        }
        #endif
        return 0
    }()

    /// `-mockJitterMS N` (DEBUG only): each delivery is late by a random 0…N ms, with the
    /// batches that fell due meanwhile landing together — the real radio's p95 300 ms /
    /// max 420 ms gaps between ~190 ms packets. Device timestamps keep their 80 Hz.
    private static let jitterMS: Int = {
        #if DEBUG
        let args = ProcessInfo.processInfo.arguments
        if let flag = args.firstIndex(of: "-mockJitterMS"), flag + 1 < args.count,
           let ms = Int(args[flag + 1]) {
            return ms
        }
        #endif
        return 0
    }()

    private var connectTask: Task<Void, Never>?
    private var connectionGeneration: UInt64 = 0
    private var pump: Task<Void, Never>?
    private var elapsedSamples: UInt64 = 0
    private var deviceMicros: UInt32 = 0
    private var tareOffsetKg: Double = 0

    init(profile: MockForceProfile = .clean) {
        self.profile = profile
    }

    deinit {
        connectTask?.cancel()
        pump?.cancel()
    }

    // MARK: - ProgressorClient

    func connect() {
        guard !state.isConnected, !state.isBusy else { return }
        connectionGeneration &+= 1
        let generation = connectionGeneration
        state = .scanning
        connectTask = Task { [weak self] in
            // A beat of latency so the connecting UI is actually exercised rather
            // than skipped past in a single frame.
            do {
                try await Task.sleep(for: .milliseconds(300))
            } catch {
                return
            }
            guard !Task.isCancelled,
                  let self,
                  self.connectionGeneration == generation else { return }
            self.state = .connecting

            do {
                try await Task.sleep(for: .milliseconds(150))
            } catch {
                return
            }
            guard !Task.isCancelled,
                  self.connectionGeneration == generation,
                  self.state == .connecting else { return }
            self.connectTask = nil
            self.state = .connected
            self.onEvent?(.appVersion("mock-1.0"))
            self.onEvent?(.battery(millivolts: 3980))
        }
    }

    func disconnect() {
        connectionGeneration &+= 1
        connectTask?.cancel()
        connectTask = nil
        stopPump()
        state = .disconnected(reason: nil)
    }

    func send(_ command: ProgressorCommand) {
        guard state.isConnected else { return }
        switch command {
        case .tare:
            // Tare zeroes whatever is on the gauge right now — the same trap as the
            // real device: tare under load and every reading after it is wrong.
            tareOffsetKg = rawForceNow()
        case .startWeightMeasurement:
            // `startStreaming(cause:)` is the only start path.
            break
        case .stopWeightMeasurement:
            stopPump()
        case .getBatteryVoltage:
            onEvent?(.battery(millivolts: 3980))
        case .getAppVersion:
            onEvent?(.appVersion("mock-1.0"))
        case .enterSleep:
            stopPump()
            state = .disconnected(reason: String(localized: "Device asleep"))
        default:
            break
        }
    }

    func startStreaming(cause: StreamStartCause) {
        guard state.isConnected else { return }
        onDiagnostic?(.streamStartWritten(cause))
        deviceMicros = 0
        startPump()
    }

    // MARK: - Sample pump

    private func startPump() {
        guard pump == nil else { return }
        elapsedSamples = 0
        deviceMicros = 0
        pump = Task { [weak self] in
            // Batches fall due on the device's own 80 Hz schedule; the pump only decides
            // WHEN it hands them over, and every batch that fell due while it slept lands
            // together, as a radio stack delivers late notifications.
            let batchMS = Int(1000 * Double(Self.batchSize) / Self.sampleHz)
            let batch = Duration.milliseconds(batchMS)
            let clump = Duration.milliseconds(max(Self.clumpMS, batchMS))
            var due = ContinuousClock.now + batch
            while !Task.isCancelled {
                let jitter = Duration.milliseconds(Self.jitterMS > 0 ? Int.random(in: 0...Self.jitterMS) : 0)
                do {
                    try await Task.sleep(until: due + (clump - batch) + jitter, clock: .continuous)
                } catch {
                    return
                }
                guard let self else { return }
                let now = ContinuousClock.now
                // Everything handed over in one wake is ONE packet, marked once: the real
                // radio's late notification carries all its samples under a single mark,
                // and the store learns the packet span from packet starts.
                var packetStart = true
                while due <= now {
                    self.emitBatch(packetStart: packetStart)
                    packetStart = false
                    due += batch
                }
            }
        }
    }

    private func stopPump() {
        pump?.cancel()
        pump = nil
        // Each demo stream replays the profile from zero. Release its synthetic load
        // here too, so the next session's pre-start tare cannot capture the last pull.
        elapsedSamples = 0
        tareOffsetKg = 0
    }

    /// One notification's worth of samples, exactly as the device batches them —
    /// which is what makes the runner's "accrue from device timestamps, not arrival
    /// time" rule testable against something realistic.
    private func emitBatch(packetStart: Bool = true) {
        onPacketBoundary?(.began(receivedAt: ProcessInfo.processInfo.systemUptime))
        defer { onPacketBoundary?(.ended) }
        for index in 0..<Self.batchSize {
            let seconds = Double(elapsedSamples) / Self.sampleHz
            let kg = MockForceProfile.force(at: seconds, profile: profile) - tareOffsetKg
            onEvent?(.sample(ForceSample(kg: kg, deviceMicros: deviceMicros,
                                         isBatchStart: packetStart && index == 0)))
            elapsedSamples += 1
            // Wrapping on purpose: over a long session the real device's UInt32 µs
            // clock rolls over at ~71.6 minutes, and the app must survive it.
            deviceMicros = deviceMicros &+ Self.microsPerSample
        }
    }

    private func rawForceNow() -> Double {
        MockForceProfile.force(at: Double(elapsedSamples) / Self.sampleHz, profile: profile)
    }
}

// MARK: - Force profiles

/// Scripted force traces. Pure functions of elapsed time, so a test can sample the
/// same curve the UI sees without running any timers.
enum MockForceProfile: String, CaseIterable, Sendable {
    /// Textbook: sharp ramp, steady plateau, clean release.
    case clean
    /// Wobbles across the threshold and briefly drops — exercises the hysteresis and
    /// RE-GRIP handling.
    case shaky
    /// Fades through the hold, the way a real set's last rep does.
    case weak
    /// Nothing on the gauge. For checking idle/zero-drift behaviour.
    case idle
    /// A critical force test done properly: all-out 7 s pulls on a 10 s cycle, decaying
    /// from about 35 kg to an 18 kg plateau, with every fifth pull held a little past the
    /// bell. The demo gauge switches to it while a test runs, so demo mode (App Review
    /// included) sees a real-looking plateau rather than the routine's 10-on/20-off shape.
    case allOut

    /// The default no-hang shape (10 s on, 20 s off), with the pull held a second and a
    /// half longer: the ramps and the engage debounce eat into a bare 10 s, so a demo rep
    /// ended on RE-GRIP with a second to go and waited out a whole rest to finish.
    static let workSeconds: Double = 11.5
    static let restSeconds: Double = 18.5

    static func force(at seconds: Double, profile: MockForceProfile) -> Double {
        let jitter = sin(seconds * 37.7) * 0.18 + sin(seconds * 13.1) * 0.1
        guard profile != .idle else { return max(0, 0.15 + jitter * 0.3) }
        if profile == .allOut { return allOutForce(at: seconds, jitter: jitter) }

        let cycle = workSeconds + restSeconds
        let phase = seconds.truncatingRemainder(dividingBy: cycle)
        guard phase < workSeconds else { return max(0, 0.2 + jitter * 0.3) }

        // Ramp on and off so the trace has real edges to detect.
        let rampIn = min(1, phase / 0.45)
        let rampOut = min(1, (workSeconds - phase) / 0.35)
        let envelope = min(rampIn, rampOut)

        let plateau: Double
        switch profile {
        case .clean:
            plateau = 22 + jitter
        case .shaky:
            // Rides around the threshold with a genuine dip near the middle.
            let wobble = sin(phase * 2.9) * 3.5
            let dip = (phase > 5.2 && phase < 5.7) ? -14.0 : 0
            plateau = 20 + wobble + dip + jitter
        case .weak:
            plateau = 24 - (phase / workSeconds) * 12 + jitter
        case .idle, .allOut:
            plateau = 0
        }
        return max(0, plateau * envelope)
    }

    private static func allOutForce(at elapsed: Double, jitter: Double) -> Double {
        // Three seconds of setting up first, so the armed PULL TO START state (and, between
        // hands, "Right hand next") is on screen before the first pull.
        let seconds = elapsed - 3
        guard seconds >= 0 else { return max(0, 0.2 + jitter * 0.3) }
        let rep = Int(seconds / 10)
        let phase = seconds - Double(rep) * 10
        let hold = rep % 5 == 2 ? 7.8 : 6.9
        guard phase < hold else { return max(0, 0.2 + jitter * 0.3) }
        let start = 18 + 17 * exp(-Double(rep) / 4.5)
        // Each pull fades within itself, as a real all-out effort does.
        let level = start * (1 - 0.14 * phase / 7) + jitter * 2
        let envelope = min(1, phase / 0.25, (hold - phase) / 0.2)
        return max(0, level * envelope)
    }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/ble/MockProgressorClient.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ble

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import run.nuri.getagrip.engine.ForceSample
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.ProgressorCommand
import run.nuri.getagrip.engine.ProgressorEvent
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

/// A synthetic Progressor — the only way to run the app without a phone and a real gauge
/// (**the emulator has no working Bluetooth**), and what demo mode uses, so someone without
/// hardware can watch a whole session.
///
/// **Always compiled in**, and reached two ways: the `--ez mockDevice true` launch extra
/// (iOS's `-mockDevice`, read by `DeviceStore`) and the "Try demo mode" button in the
/// gauge's disconnected state. A debug-only mock would strand anyone without hardware.
///
/// TRANSLATION NOTE: Swift's `Task` inherits `@MainActor`; Kotlin takes the scope
/// explicitly (the app passes `Dispatchers.Main.immediate`, a test its test scope), so it
/// is a constructor parameter.
class MockProgressorClient(
    private val scope: CoroutineScope,
    var profile: MockForceProfile = MockForceProfile.clean,
    private val clock: HostClock = SystemHostClock,
) : ProgressorClient {

    override var onEvent: ((ProgressorEvent) -> Unit)? = null
    override var onPacketBoundary: ((PacketBoundary) -> Unit)? = null
    override var onStateChange: ((ProgressorConnectionState) -> Unit)? = null
    override var onDiagnostic: ((ProgressorClientDiagnostic) -> Unit)? = null

    override var state: ProgressorConnectionState = ProgressorConnectionState.Idle
        private set(value) {
            val changed = field != value
            field = value
            if (changed) onStateChange?.invoke(value)
        }

    override val deviceName: String? = "Progressor_MOCK"

    private companion object {
        /// The device streams 80 samples/sec, delivered in batches of ~8.
        const val sampleHz: Double = 80.0
        const val batchSize: Int = 8
        val microsPerSample: UInt = (1_000_000 / 80).toUInt()
        val batteryMillivolts: UInt = 3980u
        const val firmware = "mock-1.0"
    }

    private var connectJob: Job? = null
    private var connectionGeneration: ULong = 0uL
    private var pump: Job? = null
    private var elapsedSamples: ULong = 0uL
    private var deviceMicros: UInt = 0u
    private var tareOffsetKg: Double = 0.0

    // MARK: - ProgressorClient

    override fun connect() {
        if (state.isConnected || state.isBusy) return
        connectionGeneration += 1uL
        val generation = connectionGeneration
        state = ProgressorConnectionState.Scanning
        connectJob = scope.launch {
            // A beat of latency so the connecting UI is exercised rather than skipped in
            // one frame.
            delay(300)
            if (connectionGeneration != generation) return@launch
            state = ProgressorConnectionState.Connecting

            delay(150)
            if (connectionGeneration != generation ||
                state != ProgressorConnectionState.Connecting
            ) {
                return@launch
            }
            connectJob = null
            state = ProgressorConnectionState.Connected
            onEvent?.invoke(ProgressorEvent.AppVersion(firmware))
            onEvent?.invoke(ProgressorEvent.Battery(batteryMillivolts))
        }
    }

    override fun disconnect() {
        connectionGeneration += 1uL
        connectJob?.cancel()
        connectJob = null
        stopPump()
        state = ProgressorConnectionState.Disconnected(reason = null)
    }

    override fun send(command: ProgressorCommand) {
        if (!state.isConnected) return
        when (command) {
            ProgressorCommand.tare ->
                // Tare zeroes whatever is on the gauge now — the real device's trap: tare
                // under load and every later reading is wrong.
                tareOffsetKg = rawForceNow()

            ProgressorCommand.startWeightMeasurement -> {
                // Starts need a cause; `startStreaming(cause)` is the only start path.
            }

            ProgressorCommand.stopWeightMeasurement -> stopPump()

            ProgressorCommand.getBatteryVoltage ->
                onEvent?.invoke(ProgressorEvent.Battery(batteryMillivolts))

            ProgressorCommand.getAppVersion ->
                onEvent?.invoke(ProgressorEvent.AppVersion(firmware))

            ProgressorCommand.enterSleep -> {
                stopPump()
                state = ProgressorConnectionState.Disconnected(L10n.tr("Device asleep"))
            }

            else -> Unit
        }
    }

    override fun startStreaming(cause: StreamStartCause) {
        if (!state.isConnected) return
        onDiagnostic?.invoke(ProgressorClientDiagnostic.StreamStartWritten(cause))
        deviceMicros = 0u
        startPump()
    }

    // MARK: - Sample pump

    private fun startPump() {
        if (pump != null) return
        elapsedSamples = 0uL
        deviceMicros = 0u
        pump = scope.launch {
            while (isActive) {
                emitBatch()
                delay((1000 * batchSize / sampleHz).toLong())
            }
        }
    }

    private fun stopPump() {
        pump?.cancel()
        pump = null
        // A new demo stream replays from zero; release the synthetic load first so the
        // pre-start tare cannot capture the last plateau.
        elapsedSamples = 0uL
        tareOffsetKg = 0.0
    }

    /// One notification's worth of samples, batched as the device does, so "accrue from
    /// device timestamps, not arrival time" is tested against something realistic.
    private fun emitBatch() = withPacket(clock.uptimeSeconds()) {
        for (index in 0 until batchSize) {
            val seconds = elapsedSamples.toDouble() / sampleHz
            val kg = MockForceProfile.force(seconds, profile) - tareOffsetKg
            onEvent?.invoke(
                ProgressorEvent.Sample(
                    ForceSample(kg = kg, deviceMicros = deviceMicros, isBatchStart = index == 0),
                ),
            )
            elapsedSamples += 1uL
            // Wraps on purpose: the device's UInt32 µs clock rolls over at ~71.6 minutes.
            // Kotlin unsigned arithmetic wraps natively (Swift's `&+`).
            deviceMicros += microsPerSample
        }
    }

    private fun rawForceNow(): Double =
        MockForceProfile.force(elapsedSamples.toDouble() / sampleHz, profile)
}

// MARK: - Force profiles

/// Scripted force traces as pure functions of elapsed time: tests sample the curve without
/// timers, and the same peak recurs run to run (correct, not a bug).
enum class MockForceProfile(val rawValue: String) {
    /// Textbook: sharp ramp, steady plateau, clean release.
    clean("clean"),

    /// Wobbles across the threshold and briefly drops — tests hysteresis and dropout
    /// handling.
    shaky("shaky"),

    /// Fades through the hold, the way a real set's last rep does.
    weak("weak"),

    /// Nothing on the gauge. For checking idle/zero-drift behaviour.
    idle("idle"),

    /// A critical force test done properly: all-out 7 s pulls on a 10 s cycle, decaying
    /// from about 35 kg to an 18 kg plateau, with every fifth pull held a little past the
    /// bell. The demo gauge switches to it while a test runs, so demo mode (store review
    /// included) sees a real-looking plateau rather than the routine's 10-on/20-off shape.
    allOut("allOut");

    companion object {
        /// Matches the default no-hang shape (10 s on, 20 s off) so a mock run lines up
        /// with a real routine.
        const val workSeconds: Double = 10.0
        const val restSeconds: Double = 20.0

        fun fromRaw(raw: String): MockForceProfile? = entries.firstOrNull { it.rawValue == raw }

        fun force(seconds: Double, profile: MockForceProfile): Double {
            val jitter = sin(seconds * 37.7) * 0.18 + sin(seconds * 13.1) * 0.1
            if (profile == idle) return max(0.0, 0.15 + jitter * 0.3)
            if (profile == allOut) return allOutForce(seconds, jitter)

            val cycle = workSeconds + restSeconds
            val phase = seconds % cycle
            if (phase >= workSeconds) return max(0.0, 0.2 + jitter * 0.3)

            // Ramp on and off so the trace has real edges to detect.
            val rampIn = min(1.0, phase / 0.45)
            val rampOut = min(1.0, (workSeconds - phase) / 0.35)
            val envelope = min(rampIn, rampOut)

            val plateau = when (profile) {
                clean -> 22 + jitter
                shaky -> {
                    // Rides around the threshold with a genuine dip near the middle.
                    val wobble = sin(phase * 2.9) * 3.5
                    val dip = if (phase > 5.2 && phase < 5.7) -14.0 else 0.0
                    20 + wobble + dip + jitter
                }
                weak -> 24 - (phase / workSeconds) * 12 + jitter
                idle, allOut -> 0.0
            }
            return max(0.0, plateau * envelope)
        }

        private fun allOutForce(elapsed: Double, jitter: Double): Double {
            // Three seconds of setting up first, so the armed PULL TO START state (and,
            // between hands, "Right hand next") is on screen before the first pull.
            val seconds = elapsed - 3
            if (seconds < 0) return max(0.0, 0.2 + jitter * 0.3)
            val rep = (seconds / 10).toInt()
            val phase = seconds - rep.toDouble() * 10
            val hold = if (rep % 5 == 2) 7.8 else 6.9
            if (phase >= hold) return max(0.0, 0.2 + jitter * 0.3)
            val start = 18 + 17 * exp(-rep.toDouble() / 4.5)
            // Each pull fades within itself, as a real all-out effort does.
            val level = start * (1 - 0.14 * phase / 7) + jitter * 2
            val envelope = minOf(1.0, phase / 0.25, (hold - phase) / 0.2)
            return max(0.0, level * envelope)
        }
    }
}
